### ISSUE TITLE
How to create connection to Cloud SQL using Unix Socket Domain?

### ISSUE THREAD

**Issue by hedza06** [GoogleCloudPlatform/cloud-sql-jdbc-socket-factory#1274]
### Question

I'm using Spring Boot App which is deployed on Cloud Run (GCP). The app is establishing the database connection through Cloud SQL Proxy using IAM authentication.

The goal is to implement Unix Socket Connection instead of TCP for connecting with Cloud SQL. Bellow is the `application.yml` content:
```
spring:
    cloud:
        gcp:
            sql:
                jdbc:
                    enabled: false
                database-name: db-name
                enable-iam-auth: true
                instance-connection-name: instance-name
    datasource:
        url: jdbc:postgresql:///db-name?unixSocketPath=/cloudsql/instance-name/.s.PGSQL.5432&socketFactory=com.google.cloud.sql.postgres.SocketFactory&enableIamAuth=true&sslmode=disable&cloudSqlInstance=instance-name
        username: my-username
        type: com.zaxxer.hikari.HikariDataSource
        hikari:
            maximum-pool-size: 10
```
I'm getting an error that password is not provided but I'm using the IAM authentication so the password is not needed.
More details are posted here: https://stackoverflow.com/questions/76125307/spring-boot-gcp-cloud-sql-connection

Any ideas how to configure unix socket connection with IAM auth?

### Code
org.postgresql.util.PSQLException: The server requested password-based authentication, but no password was provided by plugin null
    at org.postgresql.core.v3.AuthenticationPluginManager.lambda$withEncodedPassword$0(AuthenticationPluginManager.java:110)
    at org.postgresql.core.v3.AuthenticationPluginManager.withPassword(AuthenticationPluginManager.java:81)
    at org.postgresql.core.v3.AuthenticationPluginManager.withEncodedPassword(AuthenticationPluginManager.java:107)
    at org.postgresql.core.v3.ConnectionFactoryImpl.doAuthentication(ConnectionFactoryImpl.java:691)
    at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:180)
    at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:235)
    at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
    at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:223)
    at org.postgresql.Driver.makeConnection(Driver.java:402)
    at org.postgresql.Driver.connect(Driver.java:261)
    at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:138)
    at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:364)
    at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:206)
    at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:476)
    at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:561)
    at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:115)
    at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:112)
    at org.hibernate.engine.jdbc.connections.internal.DatasourceConnectionProviderImpl.getConnection(DatasourceConnectionProviderImpl.java:122)
    at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator$ConnectionProviderJdbcConnectionAccess.obtainConnection(JdbcEnvironmentInitiator.java:181)
    at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator.initiateService(JdbcEnvironmentInitiator.java:68)
    at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator.initiateService(JdbcEnvironmentInitiator.java:35)
    at org.hibernate.boot.registry.internal.StandardServiceRegistryImpl.initiateService(StandardServiceRegistryImpl.java:101)
    at org.hibernate.service.internal.AbstractServiceRegistryImpl.createService(AbstractServiceRegistryImpl.java:263)
    at org.hibernate.service.internal.AbstractServiceRegistryImpl.initializeService(AbstractServiceRegistryImpl.java:237)
    at org.hibernate.service.internal.AbstractServiceRegistryImpl.getService(AbstractServiceRegistryImpl.java:214)
    at org.hibernate.id.factory.internal.DefaultIdentifierGeneratorFactory.injectServices(DefaultIdentifierGeneratorFactory.java:175)
    at org.hibernate.service.internal.AbstractServiceRegistryImpl.injectDependencies(AbstractServiceRegistryImpl.java:286)
    at org.hibernate.service.internal.AbstractServiceRegistryImpl.initializeService(AbstractServiceRegistryImpl.java:243)
    at org.hibernate.service.internal.AbstractServiceRegistryImpl.getService(AbstractServiceRegistryImpl.java:214)
    at org.hibernate.boot.internal.InFlightMetadataCollectorImpl.<init>(InFlightMetadataCollectorImpl.java:173)
    at org.hibernate.boot.model.process.spi.MetadataBuildingProcess.complete(MetadataBuildingProcess.java:127)
    at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.metadata(EntityManagerFactoryBuilderImpl.java:1460)
    at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.build(EntityManagerFactoryBuilderImpl.java:1494)
    at org.springframework.orm.jpa.vendor.SpringHibernateJpaPersistenceProvider.createContainerEntityManagerFactory(SpringHibernateJpaPersistenceProvider.java:58)
    at org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean.createNativeEntityManagerFactory(LocalContainerEntityManagerFactoryBean.java:365)
    at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.buildNativeEntityManagerFactory(AbstractEntityManagerFactoryBean.java:409)
    at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.afterPropertiesSet(AbstractEntityManagerFactoryBean.java:396)
    at org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean.afterPropertiesSet(LocalContainerEntityManagerFactoryBean.java:341)
    at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1863)
    at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1800)
    at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:620)
    at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)
    at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)
    at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
    at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)
    at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)
    at org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:1154)
    at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:908)
    at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:583)
    at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:145)
    at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:745)
    at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:420)
    at org.springframework.boot.SpringApplication.run(SpringApplication.java:307)
    at org.springframework.boot.SpringApplication.run(SpringApplication.java:1317)
    at org.springframework.boot.SpringApplication.run(SpringApplication.java:1306)
    at com.glopal.ms.glfbconnector.FacebookConnectorApplication.main(FacebookConnectorApplication.java:14)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.base/java.lang.reflect.Method.invoke(Method.java:568)
    at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49)
    at org.springframework.boot.loader.Launcher.launch(Launcher.java:108)
    at org.springframework.boot.loader.Launcher.launch(Launcher.java:58)
    at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:88)

**Comment by enocom** [comment]
So disclaimer: our public docs don't make this very clear, but if you're using the Java connector, you don't need to enable the Cloud SQL integration. The Java Connector alone is sufficient to establish a connection using Auto IAM AuthN. Also, note: the built-in Cloud SQL integration doesn't support Auto IAM AuthN.

If for some reason you absolutely need a Unix socket in your connection path, we show how to do that [here](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/cloud-sql/postgres/servlet/src/main/java/com/example/cloudsql/ConnectorConnectionPoolFactory.java#L64).

**Comment by hedza06** [comment]
> So disclaimer: our public docs don't make this very clear, but if you're using the Java connector, you don't need to enable the Cloud SQL integration. The Java Connector alone is sufficient to establish a connection using Auto IAM AuthN. Also, note: the built-in Cloud SQL integration doesn't support Auto IAM AuthN.
> 
> If for some reason you absolutely need a Unix socket in your connection path, we show how to do that [here](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/cloud-sql/postgres/servlet/src/main/java/com/example/cloudsql/ConnectorConnectionPoolFactory.java#L64).

So, you're saying that it won't work if I use YAML properties and I should create configuration (DataSource) where I can define all needed properties with unixSocketPath included?

**Comment by enocom** [comment]
I see what you're saying -- no I'd expect it to work with how you're doing it. Let me look in the code to see if I'm missing something.

cc @shubha-rajan if you'd like to explain how this works.

**Comment by enocom** [comment]
Actually, I can see [here](https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/blob/main/core/src/main/java/com/google/cloud/sql/core/CoreSocketFactory.java#L182-L195) that the Java Connector + a Unix socket doesn't support Auto IAM AuthN.

**Comment by enocom** [comment]
So ideally you'd just use the Java Connector and skip the Unix socket. This would let you use Auto IAM AuthN. 

Or as another option you could configure a DataSource and subclass it like so to use Manual IAM AuthN:

``` java
public class ManualIamAuthNHikariDataSource extends HikariDataSource {

    @Override
    public String getPassword() {
        return getToken();
    }

    private String getToken() {
        // logic to get the OAuth2 token here
    }
}
```

**Comment by hedza06** [comment]
> So ideally you'd just use the Java Connector and skip the Unix socket. This would let you use Auto IAM AuthN.
> 
> Or as another option you could configure a DataSource and subclass it like so to use Manual IAM AuthN:
> 
> ```java
> public class ManualIamAuthNHikariDataSource extends HikariDataSource {
> 
>     @Override
>     public String getPassword() {
>         return getToken();
>     }
> 
>     private String getToken() {
>         // logic to get the OAuth2 token here
>     }
> }
> ```

So, what you're saying is that I need to create a configuration class which will specify connection details (url, username, instance name, unix socket path etc.) and creates data source object, and additionally, to implement a logic that fetches OAuth2 token and sets it as password which will be used in process of creating unix socket connection to the Cloud SQL?

**Comment by enocom** [comment]
Yes. If you're going to use a Unix socket, you'll have to manually provide the OAuth2 token yourself. The above approach is the cleanest one I know of.

**Comment by enocom** [comment]
Feel free to re-open if you hit any bumps.

**Comment by hedza06** [comment]
Cloud SQL Team has provided a workaround for the issue, which consists of using the latest version of the Java connector that has a fix applied for the error messages we're facing, without using a Unix socket.

Release notes can be found on this link: https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/releases

Fix is applied in version **v1.11.1.**

If you're using Spring Cloud GCP Dependency (like I am in this project) you can find release notes on this link: https://github.com/GoogleCloudPlatform/spring-cloud-gcp/releases. Versions **v3.5.1** and **v4.3.1** use the proper Java Connector with fixed issue.

This means that you don't need any custom data source configuration, just use YML props as before:
```yml
spring:
  cloud:
    gcp:
      sql:
        database-name: db-name
        enable-iam-auth: true
        instance-connection-name: instance-name
  datasource:
    hikari:
      maximum-pool-size: 10
    type: com.zaxxer.hikari.HikariDataSource
    username: usr
```

Thank you 😄
